rs/execution_environment/src/canister_manager.rs

```diff
@@ -1386,90 +1314,141 @@ impl CanisterManager {
-                    required: requested,
-                },
-                ReservationError::ReservedLimitExceed { requested, limit } => {
-                    CanisterManagerError::ReservedCyclesLimitExceededInMemoryGrow {
+                if canister.system_state.balance() < threshold + reservation_cycles {
+                    return Err(CanisterManagerError::InsufficientCyclesInMemoryGrow {
                         bytes: memory_increase,
-                        requested,
-                        limit,
-                    }
+                        available: canister.system_state.balance(),
+                        required: threshold + reservation_cycles,
+                    });
                 }
-            })?;
+
+                // Verify that the subnet has enough memory available to satisfy the
+                // requested change by the canister.
+                round_limits
+                    .subnet_available_memory
+                    .check_available_memory(memory_increase, NumBytes::from(0), NumBytes::from(0))
+                    .map_err(
+                        |_| CanisterManagerError::SubnetMemoryCapacityOverSubscribed {
+                            requested: memory_increase,
+                            available: NumBytes::from(
+                                round_limits
+                                    .subnet_available_memory
+                                    .get_execution_memory()
+                                    .max(0) as u64,
+                            ),
+                        },
+                    )?;
+
+                // Reserve needed cycles if the subnet is becoming saturated.
+                canister
+                    .system_state
+                    .can_reserve_cycles(reservation_cycles)
+                    .map_err(|err| match err {
+                        ReservationError::InsufficientCycles {
+                            requested,
+                            available,
+                        } => CanisterManagerError::InsufficientCyclesInMemoryGrow {
+                            bytes: memory_increase,
+                            available,
+                            required: requested,
+                        },
+                        ReservationError::ReservedLimitExceed { requested, limit } => {
+                            CanisterManagerError::ReservedCyclesLimitExceededInMemoryGrow {
+                                bytes: memory_increase,
+                                requested,
+                                limit,
+                            }
+                        }
+                    })?;
+            }
+        };
 
         Ok(())
     }
 
+    // IMPORTANT! This function can only be called after successful call to `self.memory_usage_checks`
+    // with the same arguments!
+    //
+    // Performs the following updates:
+    // 1. Reserve cycles on canister
+    // 2. Actually deduct memory from subnet
```

Comment:
Perhaps this is a case where we can enforce this a bit better for callers. How about we make `memory_usage_checks` return a struct that `memory_usage_updates` requires as an argument? We implement this idea for the function that validates canister settings which returns a `ValidatedSettings` struct which then ensures that other parts of the code can operate only on a validates settings set.

Reply:
Very good point - done: https://github.com/dfinity/ic/pull/4749/commits/20eca4c20417240829cb9b3829a6820fae9ef58e